pdl-compiler/src/backends/rust/serializer.rs

```diff
@@ -16,131 +16,228 @@ use crate::backends::rust::{mask_bits, types, ToIdent, ToUpperCamelCase};
 use crate::{analyzer, ast};
 use quote::{format_ident, quote};
 
-/// A single bit-field value.
+/// Generate a range check for a scalar value backed to a rust type
+/// that exceeds the actual size of the PDL field.
+fn range_check(
+    value: proc_macro2::TokenStream,
+    width: usize,
+    packet_name: &str,
+    field_name: &str,
+) -> proc_macro2::TokenStream {
+    let max_value = mask_bits(width, "u64");
+    quote! {
+        if #value > #max_value {
+            return Err(EncodeError::InvalidScalarValue {
+                packet: #packet_name,
+                field: #field_name,
+                value: #value as u64,
+                maximum_value: #max_value as u64,
+            })
+        }
+    }
+}
+
+/// Represents the computed size of a packet,
+/// compoased of constant and variable size fields.
+struct RuntimeSize {
+    constant: usize,
+    variable: Vec<proc_macro2::TokenStream>,
+}
+
+impl RuntimeSize {
+    fn payload_size() -> Self {
+        RuntimeSize { constant: 0, variable: vec![quote! { self.payload.len() }] }
+    }
+}
+
+impl std::ops::AddAssign<&RuntimeSize> for RuntimeSize {
+    fn add_assign(&mut self, other: &RuntimeSize) {
+        self.constant += other.constant;
+        self.variable.extend_from_slice(&other.variable)
+    }
+}
+
+impl quote::ToTokens for RuntimeSize {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.extend(match self {
+            RuntimeSize { constant, variable } if variable.is_empty() => quote! { #constant },
```

Comment:
I noticed that the generated code has a few `1usize` constants and I bet they are generated here. It's not really a problem and perhaps it's even nice for understanding that the code reads `123usize + foo.len()` or similar?

If you want to make things consistent, you should be able to slap on a `syn::Index::from(constant)` here and two lines below.

Reply:
Applied the suggested change :)